### DISCUSSION TITLE
Intermittent "abnormally closing connection from backend" error

### DISCUSSION THREAD

**Discussion by nguyenkims** [rspamd/rspamd#4011] [Q&A]
Hi I have noticed in Rspamd log (`/var/log/rspamd/rspamd.log`) a lot of the below error:

```
2021-12-16 09:00:11 #17951(rspamd_proxy) <cac2d4>; proxy; proxy_backend_master_error_handler: abnormally closing connection from backend: [::1]:11333, error: IO write error: Connection refused, retries left: 4
2021-12-16 09:00:11 #17951(rspamd_proxy) <cac2d4>; proxy; proxy_backend_master_error_handler: retry connection to: 127.0.0.1 retries left: 4
```

They seem to happen randomly as I can't see a pattern here. 

I have tried to increase the number of worker but the error persists. Both `/etc/rspamd/local.d/worker-proxy.inc` and `/etc/rspamd/local.d/worker-normal.inc` are set to:

```
count = 20;
```

For info, we have to handle about 400,000 emails per day. I don't have any clue at the moment for what went wrong so any idea is helpful! Thanks.

**Comment by RealSebFox** [comment]
Hello,

I found this ticket because I had a similar problem. As for now, it seems I have managed to solve it.

In the error message above, it looks as if the database is refusing connections from the local IPv6 address. I had IPv6 disabled in redis and rspamd, but still I was seeing this error. After I re-enabled IPv6 for redis and rspamd, the error was gone. Here is a reconstruction of the actions I have taken: 

In vim /etc/redis/redis.conf, I made sure ::1 is set to bind. Example: 

```
bind 127.0.0.1 ::1
```

I searched for all rspamd config files containing 127.0.0.1:

```
# grep 127.0.0.1 /etc/rspamd/local.d/*
/etc/rspamd/local.d/classifier-bayes.conf:servers = "127.0.0.1";
/etc/rspamd/local.d/redis.conf:servers = "127.0.0.1";
```

...and replaced "127.0.0.1" with "localhost" in both files:

```
servers = "localhost"
```

This may not be the perfect solution and I am not an expert. I'm just leaving this here, because this ticket is the top result when searching for this error message on the internet and the described actions have worked for me. Maybe it helps someone.

**Reply by RealSebFox** [reply]
I have found the reason why connections were made to ::1. In /etc/rspamd/rspamd.conf, `localhost:11333` is set as `bind_socket` for worker "normal", worker "controller" and worker "rspamd_proxy". In order to avoid IPv6, this can be changed (3x) to:

```
bind_socket = "127.0.0.1:11333";
```
